### PR TITLE
Add parsing to file check

### PR DESCRIPTION
**Pull request by eellison**
This allows you to embed checks in IR, making the test more readable.

E.g.
```
graph_str = 'graph(%0 : Double(5, 5)):
          # CHECK: aten::relu
          %1 : Double(5, 5) = aten::relu(%0)
          return (%1)'       
FileCheck().run(graph_str, parseIR(graph_str))
```

### CODE REVIEW

File: torch/csrc/jit/init.cpp [hunk 1]
```diff
@@ -348,11 +348,18 @@ void initJITBindings(PyObject* module) {
       },
       py::arg("qualified_name"));
 
+  m.def("parseIR", [](const std::string& input) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/init.cpp]: I think `parse_ir` is more compliant with our python style.

File: torch/csrc/jit/testing/file_check.cpp [hunk 2]
```diff
@@ -92,12 +93,22 @@ size_t assertFind(
     printQuotedString(ss, sub);
     ss << " but did not find it\n";
     found_range.highlight(ss);
-    ss << "From " << check << "\n";
+    if (extra_msg)
```
Comment (ZolotukhinM) [inline torch/csrc/jit/testing/file_check.cpp]: Nit: missing {}.

File: torch/csrc/jit/testing/file_check.cpp [hunk 3]
```diff
@@ -133,32 +156,124 @@ struct FileCheckImpl {
...
+        {CHECK_DAG, "-DAG: "},
+        {CHECK_COUNT, "-COUNT-"}, // needs special parsing
+    };
+    const std::string prefix = "# CHECK";
+
+    size_t start = 0;
+    size_t num_checks = 0;
+    start = checks_file->find(prefix, start);
+
+    while (start != std::string::npos) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/testing/file_check.cpp]: This loops seems complicated. Could we refactor it a bit? Extracting out a function for processing one line is one possible change we can make here for instance.

File: torch/csrc/jit/testing/file_check.cpp [hunk 4]
```diff
@@ -133,32 +156,124 @@ struct FileCheckImpl {
...
         last_group.push_back(check);
       } else {
         groups.push_back({check});
       }
     }
-
     has_run = false;
   }
 
+  TORCH_API void addCheck(
```
Comment (ZolotukhinM) [inline torch/csrc/jit/testing/file_check.cpp]: Looks like this change is an independent refactoring from the new feature - could you please split it into separate PRs (sorry that I didn't notice that earlier before the first split)?

File: torch/csrc/jit/testing/file_check.cpp [hunk 5]
```diff
@@ -133,32 +156,124 @@ struct FileCheckImpl {
...
+        found_match = true;
+        break;
+      }
+      if (!found_match) {
+        std::ostringstream ss;
+        ss << "Could not parse check at:\n";
+        SourceRange(checks_file, start, start + prefix.size()).highlight(ss);
+        ss << "Check for bad input.";
+        has_run = true;
+        throw std::runtime_error(ss.str());
```
Comment (ZolotukhinM) [inline torch/csrc/jit/testing/file_check.cpp]: I'm not sure we should throw every time we weren't able to parse a check tag. I can easily imagine having an input with comments "# CHECK THAT SOMETHING IS CORRECT" or something like that - we shouldn't throw an exception when we see that. I do agree that it might be helpful to throw in cases like "# CHECK-COUNT: qqq" though, maybe we should just be a bit more selective when we do that.

File: torch/csrc/jit/testing/file_check.cpp [hunk 6]
```diff
@@ -133,32 +156,124 @@ struct FileCheckImpl {
...
+      }
+
+      start = checks_file->find(prefix, start);
+      num_checks++;
+    }
+    size_t num_check_string = substringCount(checks_file, "CHECK");
```
Comment (ZolotukhinM) [inline torch/csrc/jit/testing/file_check.cpp]: Isn't this check redundant? A check in the loop already verifies that every prefix "CHECK" is followed by a valid suffix.

File: torch/csrc/jit/testing/file_check.cpp [hunk 7]
```diff
@@ -133,32 +156,124 @@ struct FileCheckImpl {
...
+    if (num_checks != 0 && num_checks == num_check_string)
+      return;
```
Comment (ZolotukhinM) [inline torch/csrc/jit/testing/file_check.cpp]: Nit: I believe it's more conventional to structure code like this:
```
if (condition)
   throw ...; // Exceptional case
// Usual case
return;
```

File: torch/csrc/jit/testing/file_check.cpp [hunk 8]
```diff
@@ -133,32 +156,124 @@ struct FileCheckImpl {
...
+    std::ostringstream ss;
+    if (num_checks < num_check_string) {
+      ss << num_checks << " checks generated but found " << num_check_string
+         << " 'CHECK' "
+         << "in input string. Check for bad input:\n"
+         << *checks_file;
+    } else if (num_checks == 0) {
+      ss << "0 checks generated. Check for bad input:\n" << *checks_file;
+    }
+    has_run = true; // no extra print after exception thrown
```
Comment (ZolotukhinM) [inline torch/csrc/jit/testing/file_check.cpp]: Do we actually need the `has_run` field? I see some value in verifying the user doesn't screw up, but at the same time I'd vote for maximum simplicity of FileCheck implementation, meaning removing non-critical checks.

**Review by ZolotukhinM — COMMENTED** [review]
Good stuff, thanks for adding this! I have some remarks inline, but overall looks good!

File: torch/csrc/jit/testing/file_check.cpp (hunk 4 above)
Comment (eellison) [reply torch/csrc/jit/testing/file_check.cpp]: imo this is part of the new feature - since there are now two ways of adding checks I had to refactor

File: torch/csrc/jit/testing/file_check.cpp (hunk 6 above)
Comment (eellison) [reply torch/csrc/jit/testing/file_check.cpp]: That will only match `# CHECK` although maybe this is overkill

File: torch/csrc/jit/testing/file_check.cpp (hunk 8 above)
Comment (eellison) [reply torch/csrc/jit/testing/file_check.cpp]: Sure i can remove it - there shouldn't be any tests cases checked in with bad inputs anyway (except for my test cases, which i can suppress output of)

File: torch/csrc/jit/testing/file_check.cpp (hunk 5 above)
Comment (eellison) [reply torch/csrc/jit/testing/file_check.cpp]: I would rather error on the side of false positive than false negatives, although I could see writing something like # CHECK ... and having an error message be annoying.

Comment (ZolotukhinM) [reply torch/csrc/jit/testing/file_check.cpp]: One other thing that might be too restrictive is that we require the there is exactly one space between '#' and 'CHECK' - maybe we could loosen that a bit too.

File: torch/csrc/jit/testing/file_check.cpp (hunk 8 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/testing/file_check.cpp]: It doesn't have to be in this PR, and if you think it's worth having it, that's fine with me too.

**Review by ZolotukhinM — APPROVED** [review]
Thanks for your comment, LGTM once the nits are fixed!
